### PR TITLE
Poll nodeip:node-port till kube-proxy updates iptables.

### PR DESCRIPTION
**Pull request by bprashanth**
Don't assume all kube proxies notice and act on the node port immediately. https://github.com/kubernetes/kubernetes/issues/18210

### CODE REVIEW

**Comment by bprashanth** [comment]
@fabioy this is causing flake and Abhi is sick today, so feel free to review if you have spare cycles

File: test/e2e/service.go [hunk 1]
```diff
@@ -40,6 +40,10 @@ import (
 	"k8s.io/kubernetes/pkg/util/wait"
 )
 
+// Maximum time a kube-proxy daemon on a node is allowed to not
+// notice a Service update, such as type=NodePort.
+const kubeProxyLagTimeout = 5 * time.Minute
```
Comment (thockin) [inline test/e2e/service.go]: 5 minutes is huge - can we start at 30 seconds?

File: test/e2e/service.go [hunk 2]
```diff
@@ -394,12 +398,17 @@ var _ = Describe("Services", func() {
 
 		By("hitting the pod through the service's NodePort")
 		ip := pickNodeIP(c)
-		testReachable(ip, nodePort)
+		// Loop for 5 minutes, because different kube-proxies might take
```
Comment (thockin) [inline test/e2e/service.go]: don't put the timeout in the comment - they will just drift

File: test/e2e/service.go [hunk 3]
```diff
@@ -470,7 +479,12 @@ var _ = Describe("Services", func() {
 		By("hitting the pod through the service's NodePort")
 		ip := pickNodeIP(f.Client)
 		nodePort1 := port.NodePort // Save for later!
-		testReachable(ip, nodePort1)
+
+		// Loop for 5 minutes, because different kube-proxies might take
```
Comment (thockin) [inline test/e2e/service.go]: same

File: test/e2e/service.go [hunk 4]
```diff
@@ -502,7 +516,13 @@ var _ = Describe("Services", func() {
 
 		By("hitting the pod through the service's NodePort")
 		ip = pickNodeIP(f.Client)
-		testReachable(ip, nodePort1)
+
+		// Loop for 5 minutes, because different kube-proxies might take
```
Comment (thockin) [inline test/e2e/service.go]: same

File: test/e2e/service.go [hunk 5]
```diff
@@ -539,7 +559,13 @@ var _ = Describe("Services", func() {
 		}
 
 		By("hitting the pod through the service's updated NodePort")
-		testReachable(ip, nodePort2)
+
+		// Loop for 5 minutes, because different kube-proxies might take
```
Comment (thockin) [inline test/e2e/service.go]: same

File: test/e2e/service.go (hunk 1 above)
Comment (bprashanth) [inline test/e2e/service.go]: 30s isn't guaranteed, but yeah. In the specify case I debugged I saw a lag of 20s. This is in part kube-proxy inefficiently handling each update of a relist of all services one by one, and in part apiserver lag.

**Comment by thockin** [comment]
just timeout and comment comments

File: test/e2e/service.go (hunk 1 above)
Comment (thockin) [inline test/e2e/service.go]: 30s lag is, to me, a pretty significant bug and if it occurs, we should fix
THAT.  I'll argue that service updates should be O(5 seconds) max, but
let's take baby steps.

Comment (bprashanth) [inline test/e2e/service.go]: Yeah it is. Compare the timestamps from kube-proxy logs in these when you have time, https://github.com/kubernetes/kubernetes/issues/18210#issuecomment-175794436. I think we should do some sort of snapshotting of iptables rules, check all these endpoints and services are in there. I haven't dug too deep into these optimizations.

**Comment by bprashanth** [comment]
Alright I gave it 45s. I think what's happening is when the watch breaks we do a relist, and kube-proxy handles each endpoints/service as a new: "ah heres a thing I should go and confirm with iptables io", type update. Unfounded theory, I need to poke around kube-proxy a bit to figure out how it gets updates better.

**Comment by fabioy** [comment]
Tests are running behind (got a green from "Jenkins unit/integration"), and change looks straightforward. Manually merging.

**Pull request merged**